src/include/txn.i

```diff
@@ -193,6 +193,166 @@ __wt_timestamp_subone(wt_timestamp_t *ts)
 
 #endif /* HAVE_TIMESTAMPS */
 
+/*
+ * __txn_op_modify_recno --
+ *      Modify the latest txn op with the given recno.
+ */
+static inline void
+__txn_op_modify_recno(WT_SESSION_IMPL *session, uint64_t recno)
+{
+	WT_TXN *txn;
+	WT_TXN_OP *op;
+
+	txn = &session->txn;
+
+	WT_ASSERT(session, txn->mod_count > 0 && recno != WT_RECNO_OOB);
+	op = txn->mod + txn->mod_count - 1;
+
+	WT_ASSERT(session, op->type == WT_TXN_OP_BASIC_COL ||
+	    op->type == WT_TXN_OP_INMEM_COL);
+
+	op->u.op_col.recno = recno;
+}
+
+/*
+ * __txn_prepared_op_resolve --
+ *      Resolve a transaction operation indirect references.
+ * In case of prepared transactions, the prepared updates could be evicted using
+ * cache overflow mechanism. Transaction operations referring these prepared
+ * updates would be referring to them using indirect references
+ * (i.e keys/recnos), which need to be resolved as part of that transaction
+ * commit/rollback.
+ */
+static inline int
+__txn_prepared_op_resolve(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit)
+{
+	WT_CURSOR *cursor;
+	WT_DECL_RET;
+	WT_TXN *txn;
+	WT_UPDATE *upd;
+	const char *open_cursor_cfg[] = {
+	    WT_CONFIG_BASE(session, WT_SESSION_open_cursor), NULL };
+
+	txn = &session->txn;
+
+	switch (op->type) {
+	case WT_TXN_OP_NONE:
+		break;
+	case WT_TXN_OP_BASIC_COL:
+	case WT_TXN_OP_BASIC_ROW:
+	case WT_TXN_OP_INMEM_COL:
+	case WT_TXN_OP_INMEM_ROW:
+		WT_RET(__wt_open_cursor(session,
+		    op->btree->dhandle->name, NULL, open_cursor_cfg, &cursor));
+
+		/*
+		 * Transaction prepare is cleared temporarily as cursor
+		 * functions are not allowed for prepared transactions.
+		 */
+		F_CLR(txn, WT_TXN_PREPARE);
+		if (op->type == WT_TXN_OP_BASIC_ROW ||
+		    op->type == WT_TXN_OP_INMEM_ROW)
+			__wt_cursor_set_raw_key(cursor, &op->u.op_row.key);
+		else
+			((WT_CURSOR_BTREE *)cursor)->iface.recno =
+			    op->u.op_col.recno;
+		F_SET(txn, WT_TXN_PREPARE);
+
+		WT_WITH_BTREE(session,
+		    op->btree, ret = __wt_btcur_search_uncommitted(
+		    (WT_CURSOR_BTREE *)cursor, &upd));
+		WT_TRET(cursor->close(cursor));
+		WT_RET(ret);
+		WT_ASSERT(session, upd != NULL);
+		op->u.op_upd = upd;
+
+		for (; upd != NULL; upd = upd->next) {
+			/* Process all the uncommitted updates of this txn. */
+			 if (upd->txnid != txn->id)
+				continue;
+
+			if (!commit) {
+				/* Rollback is just to update txn id. */
+				upd->txnid = WT_TXN_ABORTED;
+				continue;
+			}
+
+			/*
+			 * Newer updates are inserted at head of update chain,
+			 * and txn operations are added at the tail of the txn
+			 * modify chain.
+			 *
+			 * For example, a transaction has modified [k,v] as
+			 * [k,v] -> [k, u1]   (txn_op : txn_op1)
+			 * [k, u1] -> [k, u2] (txn_op : txn_op2)
+			 * update chain : u2->u1
+			 * txn_mod      : txn_op1->txn_op2.
+			 *
+			 * we save only key in the txn op, hence during
+			 * commit/rollback we cannot identify "txn_op1"
+			 * corresponds to either "u2" or "u1".
```

Comment:
Grammar:

Only the key is saved in the transaction op structure, hence we cannot identify whether "txn_op1" corresponds to "u2" or "u1" during commit/rollback.